data/shaders/pbr_ps.glsl

```diff
@@ -0,0 +1,211 @@
+// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+#version 150 core
+#include locals lights
+
+const int BASE_COLOR_MAP          = 1 << 0;
+const int NORMAL_MAP              = 1 << 1;
+const int METALLIC_ROUGHNESS_MAP  = 1 << 2;
+const int EMISSIVE_MAP            = 1 << 3;
+const int OCCLUSION_MAP           = 1 << 4;
+
+uniform sampler2D u_BaseColorSampler;
+uniform sampler2D u_NormalSampler;
+uniform sampler2D u_EmissiveSampler;
+uniform sampler2D u_MetallicRoughnessSampler;
+uniform sampler2D u_OcclusionSampler;
+
+uniform b_Globals {
+    mat4 u_ViewProj;
+    uint u_NumLights;
+};
+
+uniform b_PbrParams {
+    vec4 u_BaseColorFactor;
+    vec3 u_Camera;
+    vec3 u_EmissiveFactor;
+    vec2 u_MetallicRoughnessValues;
+    float u_NormalScale;
+    float u_OcclusionStrength;
+    int u_PbrFlags;
+};
+
+in vec3 v_Position;
+in vec2 v_TexCoord;
+in mat3 v_Tbn;
+
+out vec4 Target0;
+
+struct PbrInfo {
+    float ndotl;
+    float ndotv;
+    float ndoth;
+    float ldoth;
+    float vdoth;
+    float perceptual_roughness;
+    float metalness;
+    vec3 base_color;
+    vec3 reflectance0;
+    vec3 reflectance90;
+    float alpha_roughness;
+};
+
+const float PI = 3.141592653589793;
+const float MIN_ROUGHNESS = 0.04;
+
+float smith(float ndotv, float r)
+{
+    float tan_sq = (1.0 - ndotv * ndotv) / max((ndotv * ndotv), 0.00001);
+    return 2.0 / (1.0 + sqrt(1.0 + r * r * tan_sq));
+}
+
+float geometric_occlusion_smith_ggx(PbrInfo pbr)
+{
+    return smith(pbr.ndotl, pbr.alpha_roughness) * smith(pbr.ndotv, pbr.alpha_roughness);
+}
+
+// Basic Lambertian diffuse, implementation from Lambert's Photometria
+// https://archive.org/details/lambertsphotome00lambgoog
+vec3 lambertian_diffuse(PbrInfo pbr)
+{
+    return pbr.base_color / PI;
+}
+
+// The following equations model the Fresnel reflectance term of the spec equation
+// (aka F()) implementation of fresnel from “An Inexpensive BRDF Model for Physically
+// based Rendering” by Christophe Schlick
+vec3 fresnel_schlick(PbrInfo pbr)
+{
+    return pbr.reflectance0 + (pbr.reflectance90 - pbr.reflectance0) * pow(clamp(1.0 - pbr.vdoth, 0.0, 1.0), 5.0);
+}
+
+// The following equation(s) model the distribution of microfacet normals across
+// the area being drawn (aka D())
+// Implementation from “Average Irregularity Representation of a Roughened Surface
+// for Ray Reflection” by T. S. Trowbridge, and K. P. Reitz
+float ggx(PbrInfo pbr)
+{
+    float roughness_sq = pbr.alpha_roughness * pbr.alpha_roughness;
+    float f = (pbr.ndoth * roughness_sq - pbr.ndoth) * pbr.ndoth + 1.0;
+    return roughness_sq / (PI * f * f);
+}
+
+bool available(int flag)
+{
+    return (u_PbrFlags & flag) == flag;
```

Comment:
This probably isn't the best strategy for good parallel performance but I'm not sure what else to do.